File: LeetCode/Problems/Python/112. Path Sum.py
#Old implementation
class Solution:
    def hasPathSum(self, root: TreeNode, sum: int) -> bool:
        
        if not root:
            return False
        queue = [(root,sum)]
        
        while queue:
            curr, sum = queue.pop()
            sum -= curr.val
            if not curr.left and not curr.right and sum == 0:
                return True
            
            if curr.left:
                queue.append((curr.left, sum))
            if curr.right: 
                queue.append((curr.right, sum))
            
        return False

#DFS
#Time: O(n)
#Space: O(n)
class Solution:
    def hasPathSum(self, root: TreeNode, sum: int) -> bool:
        if not root:
            return root
        
        #dfs
        stack = [(root, sum)] 
        
        while stack:
            node, remain = stack.pop()
            if not node:
                continue
            remain -= node.val
            if not node.left and not node.right:
                if remain == 0:
                    return True
            else:
                stack.append((node.right, remain))
                stack.append((node.left, remain))
        
        return False

File: LeetCode/Problems/C++/112. Path Sum.cpp
//post order iterative
class Solution {
public:
    vector<vector<int>> pathSum(TreeNode* root, int sum) {
        vector<vector<int>> paths;
        vector<int> found_path;
        stack<TreeNode*> st;
        TreeNode* node = root;
        TreeNode* prev = NULL;
        while(node ||!st.empty()){
            if(node){
                st.push(node);
                found_path.push_back(node->val);
                sum -= node->val;
                node = node->left;  
            }
            else{
                node = st.top();
                if(node->right && prev != node->right){
                    //traverse right subtree
                    node = node->right;
                }
                else{
                    //leaf
                    if(!node->left && !node->right && sum ==0)
                        paths.emplace_back(found_path);
                    //backtracking
                    sum += node->val;
                    found_path.pop_back();
                    st.pop();
                    prev = node;
                    node = NULL;
                }
            }  
        }
        return paths;
    }
};



//recursive
class Solution {
public:
    bool hasPathSum(TreeNode* root, int sum) {
        if(!root) return false;
        
        sum-=root->val;
        if(sum==0 && !root->left && !root->right){
            return true;
        }
        return hasPathSum(root->left, sum) || hasPathSum(root->right, sum);
    }
};


//iterative
class Solution {
public:
    bool hasPathSum(TreeNode* root, int sum) {
        if(!root) return false;
        stack<pair<TreeNode*, int>> s;
        s.push({root,sum});
        
        while(!s.empty()){
            TreeNode* node = s.top().first;
            int curr_sum = s.top().second;
            s.pop();
            if(node==NULL){
                continue;
            }
            curr_sum-= node->val;
            
            if(node->left == NULL && node->right == NULL){
                if(curr_sum == 0) return true;
            }
            
            s.push({node->left, curr_sum});
            s.push({node->right, curr_sum});
            
        }
        
       return false;
    }
};